File: web/src/lib/adminColumns.js
/** Nombres canónicos del export Gravity y alias (p. ej. «Institución o Colegio»). */

export const COL_INST = 'Institución o colegio'
export const COL_GRADO = 'Grado'
export const COL_GRUPO = 'Indica el número o letra del grado'

function normKey(name) {
  return String(name ?? '')
    .trim()
    .toLowerCase()
    .normalize('NFD')
    .replace(/\p{M}/gu, '')
}

const ALIASES = new Map([
  [normKey(COL_INST), COL_INST],
  [normKey('Institución o Colegio'), COL_INST],
  [normKey(COL_GRADO), COL_GRADO],
  [normKey(COL_GRUPO), COL_GRUPO],
])

export function canonicalColumnName(name) {
  const key = normKey(name)
  return ALIASES.get(key) ?? String(name ?? '').trim()
}

/**
 * @param {{ columns: string[], rows: Record<string, unknown>[] }} sheet
 */
export function normalizeSheetColumns(sheet) {
  const columns = sheet.columns.map((c) => canonicalColumnName(c))
  const rows = sheet.rows.map((row) => {
    const out = {}
    for (const col of sheet.columns) {
      const canon = canonicalColumnName(col)
      out[canon] = row[col]
    }
    return out
  })
  return { columns, rows }
}

File: admin_columns.py
"""Nombres canónicos de columnas del export Gravity / administrador y alias habituales."""

from __future__ import annotations

import unicodedata

import pandas as pd

INST_COL = "Institución o colegio"
GRADO_COL = "Grado"
GRUPO_COL = "Indica el número o letra del grado"

REQUIRED_ADMIN_COLS = (INST_COL, GRADO_COL, GRUPO_COL)


def _norm_key(name: str) -> str:
    s = unicodedata.normalize("NFC", str(name).strip().lower())
    s = "".join(c for c in unicodedata.normalize("NFD", s) if unicodedata.category(c) != "Mn")
    return s


_ALIASES: dict[str, str] = {
    _norm_key(INST_COL): INST_COL,
    _norm_key("Institución o Colegio"): INST_COL,
    _norm_key(GRADO_COL): GRADO_COL,
    _norm_key(GRUPO_COL): GRUPO_COL,
}


def canonical_column_name(name: str) -> str:
    key = _norm_key(name)
    return _ALIASES.get(key, str(name).strip())


def normalize_admin_dataframe_columns(df: pd.DataFrame) -> pd.DataFrame:
    rename = {c: canonical_column_name(c) for c in df.columns}
    return df.rename(columns=rename)
